cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,335 @@ namespace rapidsmpf::shuffler::detail {
 
-    /// If known, the size of the GPU data buffer (in bytes).
-    std::size_t const gpu_data_size;
+    /**
+     * @brief Whether the i-th message is a control message.
+     *
+     * @param i The index of the message.
+     * @return True if the message is a control message, false otherwise.
+     */
+    inline bool is_control_message(size_t i) const {
+        return expected_num_chunks(i) > 0;
+    }
 
-    /// Metadata of the packed `cudf::table` associated with this chunk.
-    std::unique_ptr<std::vector<uint8_t>> metadata;
+    /**
+     * @brief Get the data of the i-th message, as a new ChunkBatch.
+     *
+     * @param new_chunk_id The ID of the new chunk.
+     * @param i The index of the message.
+     * @param stream The CUDA stream.
+     * @return A new ChunkBatch containing the data of the i-th message.
+     * @note This will create a copy of the packed data. If there is only one message and
+     * the message is a data message, the buffers will be moved to the new ChunkBatch.
+     * Otherwise a new ChunkBatch will be created by copying data.
+     *
+     * @throws std::out_of_range if the index is out of bounds.
+     */
+    Chunk get_data(ChunkID new_chunk_id, size_t i, rmm::cuda_stream_view stream);
 
-    /// GPU data buffer of the packed `cudf::table` associated with this chunk.
-    std::unique_ptr<Buffer> gpu_data;
+    /**
+     * @brief Get the size of the metadata of the i-th message.
+     *
+     * @param i The index of the message.
+     * @return The size of the metadata of the message. Zero when the message is a
+     * control message, otherwise the size of `PackedData::metadata`.
+     */
+    inline uint32_t metadata_size(size_t i) const {
+        assert(!meta_offsets_.empty());
+        assert(!is_control_message(i));
+        return i == 0 ? meta_offsets_.at(0)
+                      : meta_offsets_.at(i) - meta_offsets_.at(i - 1);
+    }
 
     /**
-     * @brief Construct a new chunk of a partition.
+     * @brief Get the size of the packed data of the i-th message.
      *
-     * @param pid The ID of the partition this chunk is part of.
-     * @param cid The ID of the chunk.
-     * @param gpu_data_size If known, the size of the gpu data buffer (in bytes).
-     * @param metadata The metadata of the packed `cudf::table` that makes up this
-     * chunk.
-     *  @param gpu_data The gpu_data of the packed `cudf::table` that makes up this
-     * chunk.
+     * @param i The index of the message.
+     * @return The size of the packed data of the message. Zero when the message is a
+     * control message, otherwise the size of `PackedData::gpu_data` of the message.
      */
-    Chunk(
-        PartID pid,
-        ChunkID cid,
-        std::size_t gpu_data_size,
-        std::unique_ptr<std::vector<uint8_t>> metadata,
-        std::unique_ptr<Buffer> gpu_data
-    );
+    inline size_t data_size(size_t i) const {
+        assert(!data_offsets_.empty());
+        assert(!is_control_message(i));
```

Comment:
*question*: Should we use `RAPIDSMPF_EXPECTS` here?

Reply:
I was previously thinking about only populate `data_offsets_` only when there are data messages. But ended up populating `data_offsets_` and `meta_offsets_` for all messages. So, we can remove these asserts.